ttnn/cpp/ttnn/operations/pool/generic/device/pool_multi_core_program_factory.cpp

```diff
@@ -1,19 +1,97 @@
-// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
-//
+// SPDX-FileCopyrightText: © 2024 Tenstorrent AI ULC
 // SPDX-License-Identifier: Apache-2.0
 
 #include "pool_op.hpp"
 #include "tt-metalium/circular_buffer.hpp"
 #include "tt-metalium/circular_buffer_config.hpp"
 #include "ttnn/operations/cb_utils.hpp"
 #include "ttnn/operations/pool/pool_utils.hpp"
+#include "tt-metalium/host_buffer.hpp"
+#include "tt-metalium/buffer.hpp"
+#include <vector>
+#include "ttnn/tensor/storage.hpp"
 #include <tt-metalium/hal.hpp>
+#include <algorithm>
 
 namespace ttnn::operations::pool {
 /**
  * Generic pool implementation that uses the new sliding window infrastructure.
  */
 
+// This function creates a scalar config tensor for the AvgPool2D operation. It is entirely made of
+// a vector of ScalarInfo structs, which are used to configure the pooling operation. The config tensor
+// is filled with out_nhw_per_core number of ScalarInfos for each core and then sharded across the cores.
+// Since we don't usually have that many different scalars, we fill the rest of the config tensor with 0s.
+static Tensor create_scalar_config_tensor(
+    AvgPoolConfig config, uint32_t in_n, uint32_t num_shards_c, CoreRangeSet all_cores) {
+    auto ranges = all_cores.ranges();
+
+    std::vector<uint32_t> config_vector;
+
+    uint32_t output_stick_w = 0;
+    uint32_t output_stick_h = 0;
+    uint32_t output_stick_c = 0;
+    uint32_t output_stick_n = 0;
+
+    uint32_t max_scalars_cnt = 0;
+    std::vector<std::vector<ScalarInfo>> scalars_per_core = {};
+
+    for (const auto& range : ranges) {
+        for (auto it = range.begin(); it != range.end(); ++it) {
+            std::vector<ScalarInfo> scalars = get_bf16_avg_pool_config_scalars(config, output_stick_w, output_stick_h);
+
+            max_scalars_cnt = std::max(max_scalars_cnt, (uint32_t)scalars.size());
+            scalars_per_core.push_back(scalars);
```

Comment:
maybe you can avoid to copy of the vector here, you can just 
scalars_per_core.emplace_back(get_bf16_avg_pool_config_scalars(...));
max_scalars_cnt = std::max(max_scalars_cnt, scalars_per_core.back.size());
you can also use size_t for max_scalars_cnt to avoid casts